sc-memory/sc-memory/sc_event_wait.hpp

```diff
@@ -0,0 +1,186 @@
+   */
+  _SC_EXTERN ScEventWaiter(ScMemoryContext const & context, ScAddr const & subscriptionElementAddr) noexcept;
+
+  /*!
+   * @brief Constructor for ScEventWaiter.
+   * @param context A sc-memory context used to wait sc-event.
+   * @param eventClassAddr A type of sc-event.
+   * @param subscriptionElementAddr An address of sc-element to wait sc-event for.
+   */
+  _SC_EXTERN ScEventWaiter(
+      ScMemoryContext const & context,
+      ScAddr const & eventClassAddr,
+      ScAddr const & subscriptionElementAddr) noexcept;
+
+  virtual bool OnEvent(TScEvent const & event);
+
+private:
+  ScElementaryEventSubscription<TScEvent> m_eventSubscription;
+};
+
+/*!
+ * @class ScConditionWaiter
+ * @brief Implements waiting for a condition based on an event.
+ * @tparam TScEvent Type of event.
+ */
+template <class TScEvent>
+class _SC_EXTERN ScConditionWaiter final : public ScEventWaiter<TScEvent>
+{
+  static_assert(std::is_base_of<ScEvent, TScEvent>::value, "TScEvent type must be derived from ScEvent type.");
+
+  friend class ScAgentContext;
+
+protected:
+  using DelegateCheckFunc = std::function<bool(TScEvent const &)>;
+
+  /*!
+   * @brief Constructor for ScConditionWaiter.
+   * @param context A sc-memory context used to wait sc-event with condition.
+   * @param subscriptionElementAddr An address of sc-element to wait event for.
+   * @param func Delegate function to check the condition.
+   */
+  _SC_EXTERN ScConditionWaiter(
+      ScMemoryContext const & context,
+      ScAddr const & subscriptionElementAddr,
+      DelegateCheckFunc const & func) noexcept;
+
+  /*!
+   * @brief Constructor for ScConditionWaiter.
+   * @param context A sc-memory context used to wait sc-event with condition.
+   * @param eventClassAddr A type of sc-event.
+   * @param subscriptionElementAddr An address of sc-element to wait event for.
+   * @param func Delegate function to check the condition.
+   */
+  _SC_EXTERN ScConditionWaiter(
+      ScMemoryContext const & context,
+      ScAddr const & eventClassAddr,
+      ScAddr const & subscriptionElementAddr,
+      DelegateCheckFunc const & func) noexcept;
+
+private:
+  bool OnEvent(TScEvent const & event) override;
+
+private:
+  DelegateCheckFunc m_checkFunc;
+};
+
+/*!
+ * @class ScWaiterActionFinished
+ * @brief Implements waiting for an action to finish.
+ */
+class _SC_EXTERN ScWaiterActionFinished final
+  : public ScEventWaiter<ScEventAfterGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>>
+{
+  friend class ScAction;
+
+protected:
+  /*!
+   * @brief Constructor for ScWaiterActionFinished.
+   * @param context A sc-memory context used to wait while action will be finished.
+   * @param actionAddr An address of the action.
+   */
+  _SC_EXTERN ScWaiterActionFinished(ScMemoryContext const & context, ScAddr const & actionAddr);
+
+private:
+  bool OnEvent(ScEventAfterGenerateIncomingArc<ScType::EdgeAccessConstPosPerm> const & event) override;
+};
```

Comment:
why can't I use this waiter using agentContext? maybe it's better to allow users to use official action waiter then to let them come up with their own implementations that may be not so optimized?